File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int lowerBound(vector<int>& nums, int target) {
        int s = 0;
        int e = nums.size()-1;
        
        int ans = -1;
        
        while(s <= e) {
            int mid = s + (e-s)/2;  // avoids integer overflow
            if(nums[mid] == target) {
                // don't return ans now and continue to search on the left side 
                // update e to check for more occurence of target
                ans = mid;
                e = mid-1;  
            }
            else if(nums[mid] < target) {
                s = mid + 1;
            }
            else {
                e = mid - 1;
            }
        }
        
        return ans;
    }
    int upperBound(vector<int>& nums, int target) {
        int s = 0;
        int e = nums.size()-1;
        
        int ans = -1;
        
        while(s <= e) {
            int mid = s + (e-s)/2;  // avoids integer overflow
            if(nums[mid] == target) {
               // don't return ans now and continue to search on the right side 
                // update s to check for more occurence of target
                ans = mid;
                s = mid+1;  
            }
            else if(nums[mid] < target) {
                s = mid + 1;
            }
            else {
                e = mid - 1;
            }
        }
        
        return ans;
    }
    // Pure BS
    vector<int> searchRange(vector<int>& nums, int target) {
        int lb = lowerBound(nums, target);
        int ub = upperBound(nums, target);
        return {lb, ub};
    }
    
    
    // BS + Linear
    /*
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans = {-1,-1};

        int s = 0, e = nums.size()-1; 
        while(s <= e) {
            int mid = (s+e)/2;
            if(nums[mid] == target) {
                int first = mid;
                int last = mid;
                while(first >= 0 && nums[first] == target) {
                    first--;
                }
                while(last < nums.size() && nums[last] == target) {
                    last++;
                }
                ans[0] = first+1;
                ans[1] = last-1;
                break;
            }
            else if(nums[mid] > target) {
                e = mid - 1;
            }
            else {
                s = mid + 1;
            }
        }

        return ans;
    }*/
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int ceil = findCeil(nums, target);
        int floor = findFloor(nums, target);
        
        if(ceil == -1 || floor == -1) {
            return new int[] {-1, -1};
        }
        return new int[] {floor, ceil};
        
    }
    int findCeil(int[] nums, int target) {
        int l = 0;
        int r = nums.length - 1;
        int ceil = -1;
        
        while(l <= r) {
            int mid = l + (r - l)/2;
            
            if(nums[mid] == target) {
                l = mid + 1;
                ceil = mid;
            }
            else if(nums[mid] < target) {
                l = mid + 1;
            }
            else {
                r = mid - 1;
            }
        }
        return ceil; 
        
    }
    int findFloor(int[] nums, int target) {
        int l = 0;
        int r = nums.length - 1;
        int floor = -1;
        
        while(l <= r) {
            int mid = l + (r - l)/2;
            
            if(nums[mid] == target) {
                r = mid - 1;
                floor = mid;
            }
            else if(nums[mid] < target) {
                l = mid + 1;
            }
            else {
                r = mid - 1;
            }
        }
        return floor;
        
        
    }
}